### ISSUE TITLE
COW cp (--reflink) doesn't work across different datasets: Invalid cross-device link

### ISSUE THREAD

**Issue by darkbasic** [openzfs/zfs#15345]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| Arch Linux
Distribution Version	|
Kernel Version	| 6.6.0-rc4
Architecture	| amd64
OpenZFS Version	| git branch zfs-2.2-release + 6.6 compatibility patches

### Describe the problem you're observing

Reflinking doesn't work across different datasets.
Since https://lore.kernel.org/linux-btrfs/cover.1645194730.git.josef@toxicpanda.com/T/#mf251325026fe2e15ed5119856bf654ba4f0d298b btrfs allows to reflink across different subvolumes, so it should be possible to achieve something similar in Linux with zfs.
Not being able to reflink across different datasets vastly reduce the utility of reflinking.

### Describe how to reproduce the problem

`cp -a --reflink=always /path/to/first/dataset/file /path/to/second/dataset/`

### Include any warning/errors/backtraces from the system logs

```
[niko@arch-phoenix ~]$ cp --reflink=always ~/.cache/yay/chromium-wayland-vaapi/chromium-117.0.5938.132.tar.xz .
cp: failed to clone './chromium-117.0.5938.132.tar.xz' from '/home/niko/.cache/yay/chromium-wayland-vaapi/chromium-117.0.5938.132.tar.xz': Invalid cross-device link
```

P.S.
I have been told that `--reflink=auto` should be able to clone blocks across different datasets, but this isn't the case:

```
[niko@arch-phoenix ~]$ time cp ~/.cache/yay/chromium-wayland-vaapi/chromium-117.0.5938.132.tar.xz . && rm chromium-117.0.5938.132.tar.xz

real	0m3.136s
user	0m0.000s
sys	0m2.852s
```

```
[niko@arch-phoenix .cache]$ time cp yay/chromium-wayland-vaapi/chromium-117.0.5938.132.tar.xz . && rm chromium-117.0.5938.132.tar.xz

real	0m0.127s
user	0m0.000s
sys	0m0.127s
```

It took 3 seconds compared to 0.1 seconds when the dataset was the same, suggesting that reflinking didn't work across datasets.

**Comment by rincebrain** [comment]
[See here](https://github.com/openzfs/zfs/pull/13392#issuecomment-1742172842).

**Comment by darkbasic** [comment]
@rincebrain interesting, but somehow it doesn't work according to the `time` results.

**Comment by robn** [comment]
`--reflink=always` calls `ioctl(FICLONE)`. Linux inspects this call before passing it to the filesystem, and will reject it if source and destination files are not on a filesystem with the same superblock (and before 5.18, the same mountpoint).

This is not an OpenZFS bug as such, because if Linux would pass the call down, we would quite happily service it. Working around Linux's check here is extremely difficult, if its even possible.

(the Btrfs example is not really relevant; OpenZFS and Btrfs have a fundamentally different construction and purpose).

`--reflink=auto` calls the `copy_file_range()` syscall, which in this case means "make a new file with the same contents as this existing one and I don't care how you do it". Often OpenZFS can service this with a clone, but not always (for many good reasons). If it can't, it'll fall back to a regular content copy.

The call time is not a very good indicator that a clone vs a copy was done. To tell if it was cloned you currently have to dig around with `zdb`. But in any case, that's acceptable behaviour, because `--reflink=auto` (`copy_file_range()` allows it).

Yes, this sucks. We'll keep working on it but its complicated.

**Comment by rincebrain** [comment]
btrfs doesn't trip this because they show up subvolumes as the same "mountpoint" which is why this check doesn't bite them.

**Comment by darkbasic** [comment]
The problem is not only with different datasets, but with snapshots as well:

```
[niko@arch-phoenix ~]$ cp --reflink=always /home/.zfs/snapshot/zrepl_20231004_074556_000/niko/devel/linux-mainline.tar.gz .
cp: failed to clone './linux-mainline.tar.gz' from '/home/.zfs/snapshot/zrepl_20231004_074556_000/niko/devel/linux-mainline.tar.gz': Invalid cross-device link
[niko@arch-phoenix ~]$ mount | grep home
rpool/home on /home type zfs (rw,nodev,relatime,xattr,posixacl,casesensitive)
rpool/home@zrepl_20231004_074556_000 on /home/.zfs/snapshot/zrepl_20231004_074556_000 type zfs (ro,relatime,xattr,posixacl,casesensitive)
```

Accessing snapshots create a different mountpoint which triggers the same issue.

**Comment by robn** [comment]
Filesystems and snapshots are different datasets. Or, put another way, they're different mounts, and so different superblocks, and so Linux rejects the request.

I understand that this is frustrating, but no amount of pointing it out is going to make a quick fix happen. I'm aware of four possible solutions (or shapes of solutions):

- Linux lifting the restriction
- Adding OpenZFS-specific calls that Linux doesn't know about (and so won't intercept), and then adding support for those to common tools (like `cp`)
- Adding `zfs clonefile` or similar command to do clones directly inside OpenZFS
- Significantly modify OpenZFS to use the same superblock for all mounts

I've been quietly exploring all of these options for a few weeks now. They are all difficult and/or complicated, for different reasons, and I also have very little time available to look at it. If you've got some other idea, I'm happy to hear it.

**Comment by darkbasic** [comment]
My only suggestion is to reach Kent Overstreet @koverstreet and ask him what are his plans for bcachefs regarding this.
Lifting the Linux restriction would obviously be the best course of action and working with a (soon-to-be) mainline filesystem would be much easier.

**Comment by darkbasic** [comment]
The reason why `--reflink=auto`  (which in turn calls the `copy_file_range()` syscall) doesn't create a clone is encryption. We can't clone encrypted blocks across datasets because the key material is partially bound to the source dataset (actually its encryption root). https://github.com/openzfs/zfs/pull/14705 has a start on this.

**Comment by robn** [comment]
If you're going to take something I wrote and pass it off as your own you should at least adjust it to match the context. That's why it didn't work in your case. It can and does create clones in many other situations.

In any case, there's no bug here. These limitations are well understood and will be worked on as time and interest allows.

**Comment by darkbasic** [comment]
> If you're going to take something I wrote and pass it off as your own you should at least adjust it to match the context.

That was not my intention and I'm sorry if it felt that way. I'm just trying my best to juggle the relevant information across the two threads so that anybody who stumbles upon either of them will understand what's going on.

> That's why it didn't work in your case

I thought that was clear enough, but I've edited my past message to change "does" with "might".

**Comment by oromenahar** [comment]
> Linux lifting the restriction

I think getting a patch into the linux kernel to lift the cross device link restriction can be pretty hard. There is no usecase in the kernel and no kernel (fs)driver wich need this. I don't think that the linux kernel will accept for example a new ioctl-flag `FICLONEC` for cross device links without any usecase inside the kernel and I think the existing flags won't be changed as well, because cross reflinking over different datasets doesn't make sense. Don't know for sure but I quess so.

**Comment by darkbasic** [comment]
Unfortunately even bcachefs won't be a suitable candide because it already supports reflinking across different subvolumes.

**Comment by jittygitty** [comment]
@darkbasic Perhaps, but regardless, Kent's very smart and knowledgeable and I think nice enough to give some objective and thoughtful opinions that may be helpful.

**Comment by darkbasic** [comment]
> Perhaps, but regardless, Kent's very smart and knowledgeable and I think nice enough to give some objective and thoughtful opinions that may be helpful.

I did write to him and I linked this issue, but he simply replied that bcachefs can already reflink between subvolumes.
I guess he's pretty busy with his own stuff.

> From the user point of view, it makes lots of sense, like reorganizing data (files) between datasets.

Definitely.

**Comment by jittygitty** [comment]
@darkbasic Yea, I read some of the Linux kernel dev emails and noticed many jumping on Kent as he was trying to get bcachefs upstream, seemed very "tense and stressful" for him. (skipping linking painful emails)

Interesting quote from Kent:
"Right now what I'm hearing, in particular from Redhat, is that they want it upstream in order to commit more resources. Which, I know, is not what kernel people want to hear, but it's the chicken-and-the-egg situation I'm in." (source https://lore.kernel.org/lkml/20230706173819.36c67pf42ba4gmv4@moria.home.lan/ )

Anyway, was really great to see we got a Halloween "present" and looks like Bcachefs was finally merged into the 6.7 kernel by Linus at the end of October!
So now Linux has OCFS2, Btrfs, XFS, Bcachefs, and ZFS all supporting reflinks. 
Will be interesting to do feature/performance benchmark comparisons as well as seeing how the various filesystems do when put on a zfs ZVOL. I wish that ZVOLs had gotten a bit more love, but I digress...

Have any parts of ZFS gotten "rusty" at all over the years?? ( https://lwn.net/Articles/934692/ )

**Comment by kimono-koans** [comment]
+1

FWIW I have a use case re: [httm](https://github.com/kimono-koans/httm).  I [implemented FICLONE](https://github.com/kimono-koans/httm/pull/89) for `httm` specifically to make use of this feature with respect to ZFS.

**Comment by EchterAgo** [comment]
@darkbasic I'm seeing the same on Ubuntu with coreutils 9.1 installed, `--reflink=auto` does not attempt to call `copy_file_range`. I am able to make it successfully reflink accross datasets by specifying `--sparse=never --reflink=auto`.

**Comment by scineram** [comment]
Coreutils should have an `--reflink=zfs` option that would simply call the zfs internal clone function and have zfs do its own internal checks.

**Comment by EchterAgo** [comment]
For some reason the `--sparse=auto` detection in `coreutils` 9.1 fails for me, resulting in `cp` always trying a sparse copy unless I specify `--sparse=never`. When doing sparse copies `cp` does not use `copy_file_range`. I'm currently trying to build `coreutils` 9.4 to see if things changed.

**Comment by EchterAgo** [comment]
From trying to copy a 1GiB file:

```
lseek(3, 0, SEEK_DATA)                  = 0
fadvise64(3, 0, 0, POSIX_FADV_SEQUENTIAL) = 0
lseek(3, 0, SEEK_HOLE)                  = 1073741824
lseek(3, 0, SEEK_SET)                   = 0
```

Seems to me that would indicate there is no hole in the source file, so why does `cp` still treat it as such?

**Comment by EchterAgo** [comment]
Newer `coreutils` versions also added a `--debug` switch for `cp` which might be helpful for diagnosing this: https://github.com/coreutils/coreutils/commit/d899f9e3320bb2a4727ca894163f6b104118c973

**Comment by EchterAgo** [comment]
Ah https://github.com/coreutils/coreutils/commit/879d2180d6b58e7a83312681fbce9a1e841c2ae4

But https://github.com/coreutils/coreutils/commit/4f92de58220226e4a2ddf3475bacee2bae7f0e1d

**Comment by darkbasic** [comment]
> I am able to make it successfully reflink accross datasets by specifying --sparse=never --reflink=auto

You don't use native encryption, right? Otherwise I'm a little bit puzzled because it's **not** supposed to work with encryption.

P.S. You should disable block cloning altogether until things get figured out because of corruption reports on Gentoo.

**Comment by EchterAgo** [comment]
> > I am able to make it successfully reflink accross datasets by specifying --sparse=never --reflink=auto
> 
> You don't use native encryption, right? Otherwise I'm a little bit puzzled because it's **not** supposed to work with encryption.
> 
> P.S. You should disable block cloning altogether until things get figured out because of corruption reports on Gentoo.

No encryption but LZ4 compression. I mean reflinking works if I pass the right options, it is the sparse detection that seems to fail in coreutils 9.1. I created the source file using `dd if=/dev/urandom of=/path/file bs=1M count=1024`, so it is definitely not a sparse file, yet `cp` treats it as such.

I know about the data corruption issue, thanks.

**Comment by strugee** [comment]
> I know about the data corruption issue, thanks.

Does anyone have an issue number for those of us who don't?

**Comment by darkbasic** [comment]
Here it is: https://github.com/openzfs/zfs/issues/15526

**Comment by mschiff** [comment]
>     * Linux lifting the restriction
>     * Adding OpenZFS-specific calls that Linux doesn't know about (and so won't intercept), and then adding support for those to common tools (like `cp`)
>     * Adding `zfs clonefile` or similar command to do clones directly inside OpenZFS
>     * Significantly modify OpenZFS to use the same superblock for all mounts
> 

I think `zfs clonefile` would be a good start. I have one more idea:

* `zfs-clonefile` being its own binary which might be symlinked as `cp` and `mv` and might mimic behavior of the linked command when being called with `--reflink` 

or 

* maybe it could fork `cp` or `mv` and then 'capture' relevant syscalls and turn them into zfs internal syscalls that will create clones successfully? But I am not sure this would be even possible.

**Comment by darkbasic** [comment]
What's the status of this with https://github.com/openzfs/zfs/pull/15544 merged?

**Comment by oromenahar** [comment]
Yes, that PR is merged.

**Comment by darkbasic** [comment]
Yeah but is this fixed or do we still need more work in order to close this issue? I didn't try git master yet since #15544 got merged.

**Comment by amotin** [comment]
I think we are doing everything reasonably possible within existing Linux API limitations.  The only thing we could do more is as proposed by @robn at the beginning is add custom IOCTL and either some custom copy tool like `zfs clonefile`, or make `coreutils` to support it, but both look dirty to me.  So unless somebody wish to work in this direction I would close this issue.

**Comment by waltmck** [comment]
I'm using ZFS not-in-production and would be willing to patch the kernel. Would the following patch be safe to make `ioctl(FICLONE)` reflink copy between ZFS datasets without breaking any other filesystems?
```diff
diff --git i/fs/remap_range.c w/fs/remap_range.c
index 26afbbbfb10c..39e7d887a1dc 100644
--- i/fs/remap_range.c
+++ w/fs/remap_range.c
@@ -381,9 +381,6 @@ loff_t vfs_clone_file_range(struct file *file_in, loff_t pos_in,
 
        WARN_ON_ONCE(remap_flags & REMAP_FILE_DEDUP);
 
-       if (file_inode(file_in)->i_sb != file_inode(file_out)->i_sb)
-               return -EXDEV;
-
        ret = generic_file_rw_checks(file_in, file_out);
        if (ret < 0)
                return ret;
```

For `copy_file_range()` it seems like I would want
```patch
diff --git i/fs/read_write.c w/fs/read_write.c
index bb0ed26a0b3a..90743ad5d408 100644
--- i/fs/read_write.c
+++ w/fs/read_write.c
@@ -1585,7 +1585,7 @@ ssize_t vfs_copy_file_range(struct file *file_in, loff_t pos_in,
                ret = file_out->f_op->copy_file_range(file_in, pos_in,
                                                      file_out, pos_out,
                                                      len, flags);
-       } else if (!splice && file_in->f_op->remap_file_range && samesb) {
+       } else if (!splice && file_in->f_op->remap_file_range) {
                ret = file_in->f_op->remap_file_range(file_in, pos_in,
                                file_out, pos_out,
                                min_t(loff_t, MAX_RW_COUNT, len),
```

but again I don't know if this is safe---could it cause copies from other filesystems to fail without falling back to a non-reflink copy?

**Comment by robn** [comment]
`copy_file_range` already does the right thing; the kernel has a carve-out for NFS and CIFS, which we take advantage of.

Your `vfs_clone_file_range` change will "work", but I can't really speak to other filesystems. It depends how much they rely on that to be true. You could at least attempt to prevent clones across different types of filesystems by comparing `f_op->remap_file_range` on both files to make sure they're at least the same filesystem code under there. OpenZFS will do the right thing, but if other filesystems aren't prepared for it, who knows what happens.

**Comment by darkbasic** [comment]
I'm giving this another try with zfs native encryption, kernel 6.14.10-arch1-1, zfs-2.3.2.r56.g4c8425715c-1 and coreutils 9.7-1:

```
time cp --reflink=auto --debug /dataset1/bigfile.img /dataset2/
'/dataset1/bigfile.img' -> '/dataset2/bigfile.img'
copy offload: yes, reflink: unsupported, sparse detection: SEEK_HOLE

real	1m36.361s
user	0m0.010s
sys	1m20.056s
```

Unfortunately `--reflink=auto` still doesn't work: any idea why?

**Comment by TheJulianJES** [comment]
The encryption initialization vector for both datasets is different. A workaround to get two datasets with the same IV would be to clone (a snapshot of) `dataset1` as `dataset2`, then delete all data from `dataset2`.
That way, you have two datasets with the same IV, allowing block cloning across the datasets. I've confirmed this works, but it's obviously not a nice solution.

Also note that the data on both datasets is encrypted with the same "key" then, so if you change the user facing key of just one dataset, you'll likely still be able to get to the data of the other dataset if you are able to decrypt either dataset.
If you just use ZFS encryption to protect against drives being stolen or if you need to send the drives to manufacturers for RMA, it's likely not an issue, as they shouldn't have any keys.

I do kinda wish there would be a way we could create datasets with the same IV/underlying key to allow for this use case, as I'd rather have block cloning than have all datasets encrypted differently, but I also see why this might not be implemented.

**Comment by robn** [comment]
> `copy offload: yes, reflink: unsupported, sparse detection: SEEK_HOLE`
> Unfortunately --reflink=auto still doesn't work: any idea why?

@darkbasic this is working exactly as expected: in coreutils `cp`, `--reflink=auto` is "try `ioctl(FICLONE)`, if that fails, use `copy_file_range()`". The output is `cp`'s interpretation of what it did: `reflink: unsupported` is "`FICLONE` didn't work", `copy offload: yes` is "`copy_file_range()` succeeded."

**Comment by darkbasic** [comment]
@robn but it doesn't look like copy_file_range() ended up using clone either:

```
sudo zdb -dddd rpool/home/niko/devel 15860490
[sudo] password for niko: 
Dataset rpool/home/niko/devel [ZPL], ID 6462, cr_txg 40421, 173G, 7066365 objects, rootbp DVA[0]=<0:8792326000:1000> DVA[1]=<0:2aa85daa000:1000> [L0 DMU objset] fletcher4 uncompressed authenticated LE contiguous unique double size=1000L/1000P birth=9918899L/9918899P fill=7066365 cksum=000000296195449d:000078c5bcf024a8:00b95cf4a8f54ad4:c5a0d987a059e2a2

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
  15860490    3   128K   128K  14.5G      1K   100G   51.50  ZFS plain file
                                               176   bonus  System attributes
	dnode flags: USED_BYTES USERUSED_ACCOUNTED USEROBJUSED_ACCOUNTED 
	dnode maxblkid: 819199
		(bonus encrypted)
		(object encrypted)
```

`rpool/home/niko/devel` is the destination dataset and `15860490` is the output of `ls -i /path/filename`

@TheJulianJES what about the simpler use case of copying from a snapshot of the same dataset?

```
$ time cp --reflink=auto ~/devel/.zfs/snapshot/zrepl_20250609_073558_000/ubuntu-24.10-desktop-amd64.iso ~/devel/

real	0m0.218s
user	0m0.000s
sys	0m0.201s

$ ls -i ~/devel/ubuntu-24.10-desktop-amd64.iso 
15860494 /home/niko/devel/ubuntu-24.10-desktop-amd64.iso

$ sudo zdb -dddd rpool/home/niko/devel 15860494
Dataset rpool/home/niko/devel [ZPL], ID 6462, cr_txg 40421, 164G, 7066365 objects, rootbp DVA[0]=<0:327aa1d2000:1000> DVA[1]=<0:9294013000:1000> [L0 DMU objset] fletcher4 uncompressed authenticated LE contiguous unique double size=1000L/1000P birth=9919602L/9919602P fill=7066365 cksum=0000002992f5ee03:0000776b5d0518ab:00b501510705b8a7:bf39f3cbdbc029b4

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
  15860494    3   128K   128K  5.26G      1K  5.28G   99.98  ZFS plain file
                                               176   bonus  System attributes
	dnode flags: USED_BYTES USERUSED_ACCOUNTED USEROBJUSED_ACCOUNTED 
	dnode maxblkid: 43224
		(bonus encrypted)
		(object encrypted)
```

At least according to the time it looks like it ended up using clone, not sure about the output of zdb though.

This is the same copy using `--reflink=never`:

```
$ time cp --reflink=never ~/devel/.zfs/snapshot/zrepl_20250609_073558_000/ubuntu-24.10-desktop-amd64.iso ~/devel/

real	0m3.220s
user	0m0.010s
sys	0m2.723s

$ ls -i ~/devel/ubuntu-24.10-desktop-amd64.iso 
15860264 /home/niko/devel/ubuntu-24.10-desktop-amd64.iso

$ sudo zdb -dddd rpool/home/niko/devel 15860264
Dataset rpool/home/niko/devel [ZPL], ID 6462, cr_txg 40421, 164G, 7066365 objects, rootbp DVA[0]=<0:2ab1edfa000:1000> DVA[1]=<0:c661cdc000:1000> [L0 DMU objset] fletcher4 uncompressed authenticated LE contiguous unique double size=1000L/1000P birth=9919688L/9919688P fill=7066365 cksum=00000029f332ecbb:0000794ab2342f4a:00b8cdfe9d8eadb2:c3f0af6d4d2d6838

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
  15860264    3   128K   128K  5.26G      1K  5.28G   99.98  ZFS plain file
                                               176   bonus  System attributes
	dnode flags: USED_BYTES USERUSED_ACCOUNTED USEROBJUSED_ACCOUNTED 
	dnode maxblkid: 43224
		(bonus encrypted)
		(object encrypted)
```

**Comment by robn** [comment]
> but it doesn't look like copy_file_range() ended up using clone either:

@darkbasic it doesn't have to. It's free to effect the copy however it wishes. The contract is that reading the ranges in the source and destination objects will yield the same data.

That said, your zdb output doesn't show enough to determine what happened anyway. You will have to dump the L0 block pointers of both objects and compare their DVAs to know for sure.

**Comment by almereyda** [comment]
@TheJulianJES regarding

> I do kinda wish there would be a way we could create datasets with the same IV/underlying key to allow for this use case, as I'd rather have block cloning than have all datasets encrypted differently, but I also see why this might not be implemented.

are you aware of these issues?

- #6824
- #12649
- #15952

The latter in https://github.com/openzfs/zfs/issues/15952#issuecomment-1974945658 also cites https://github.com/openzfs/zfs/issues/6824#issuecomment-1830273786 where it says:

> Theoretically, though, the implementation shouldn't be that hard. We just need to save the parameters needed to derive the user key to a file and then allow the proper hooks into the kernel to allow those parameters to be used instead of the ones that are saved in zfs. Then that would need to be wired out into the command line interface.

It seems noone is against this, but there's just not enough capacity to get to it soon.

**Comment by darkbasic** [comment]
About the potential solutions that were proposed (like `zfs clonefile` or patching coreutils), would they also work for datasets with different IVs?

**Comment by robn** [comment]
> About the potential solutions that were proposed (like zfs clonefile or patching coreutils), would they also work for datasets with different IVs?

No. That's all internal to OpenZFS. A encrypted block needs some key material from its containing dataset in order to be decrypted. Therefore, the block can't be put into a different dataset, because it won't have the correct key material won't match up.

It's unrelated to alternate ways to trigger cloning. It's irrelevant what we can and can't clone, and when. If the kernel won't even pass the cloning request on to us, none of it matters.

**Comment by darkbasic** [comment]
So even if the encryption key is same across the whole pool since each dataset has a different IV there is basically no hope to ever get cloning across different datasets working?

**Comment by kerberizer** [comment]
> So even if the encryption key is same across the whole pool since each dataset has a different IV there is basically no hope to ever get cloning across different datasets working?

Not sure if I understand your comment well, but, FWIW, ~~block cloning between encrypted datasets **that share the same encryption root** has been possible ever since #15544 was merged. I use it regularly without problem.~~

**Comment by robn** [comment]
Actually not the encryption root; they must be in the same "clone group": the dataset, its snapshots and any clones created from those snapshots. Those necessarily share a master key, so blocks can be cloned between them.

**Comment by darkbasic** [comment]
That's exactly what I experienced: I can clone from a snapshot (which is VERY handy) but I cannot clone to a completely unrelated dataset that shares the same encryption key.
Is the latter something that could eventually be implemented?

**Comment by kerberizer** [comment]
> Actually not the encryption root; they must be in the same "clone group"

Indeed, sorry about the noise from my side (and I just realized my "without problem" case isn't applicable here). I've stricken my comment to not confuse people who wouldn't read further.
